File: apps/api/app/schemas/decision.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


DecisionKind = Literal["decision", "notification"]
DecisionStatus = Literal["draft", "registered", "effective", "expired", "revoked", "archived"]


class DecisionBase(BaseModel):
    document_id: str
    decision_kind: DecisionKind
    document_number: str | None = Field(default=None, max_length=128)
    document_symbol: str | None = Field(default=None, max_length=128)
    issued_date: date | None = None
    issuing_agency: str | None = Field(default=None, max_length=255)
    excerpt: str | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    status: DecisionStatus = "draft"
    notes: str | None = None


class DecisionCreateRequest(DecisionBase):
    pass


class DecisionUpdateRequest(BaseModel):
    decision_kind: DecisionKind | None = None
    document_number: str | None = Field(default=None, max_length=128)
    document_symbol: str | None = Field(default=None, max_length=128)
    issued_date: date | None = None
    issuing_agency: str | None = Field(default=None, max_length=255)
    excerpt: str | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    status: DecisionStatus | None = None
    notes: str | None = None


class DecisionRead(BaseModel):
    id: str
    document_id: str
    document_title: str | None = None
    document_type: str | None = None
    document_status: str | None = None
    decision_kind: str
    document_number: str | None = None
    document_symbol: str | None = None
    issued_date: date | None = None
    issuing_agency: str | None = None
    excerpt: str | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    status: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class DecisionListResponse(BaseModel):
    items: list[DecisionRead]
    total: int
    limit: int
    offset: int

File: apps/web/types/decision.ts
export type DecisionKind = 'decision' | 'notification'
export type DecisionStatus = 'draft' | 'registered' | 'effective' | 'expired' | 'revoked' | 'archived'

export interface DecisionItem {
  id: string
  document_id: string
  document_title?: string | null
  document_type?: string | null
  document_status?: string | null
  decision_kind: DecisionKind
  document_number?: string | null
  document_symbol?: string | null
  issued_date?: string | null
  issuing_agency?: string | null
  excerpt?: string | null
  effective_from?: string | null
  effective_to?: string | null
  status: DecisionStatus
  notes?: string | null
  created_at: string
  updated_at: string
}

export interface DecisionInput {
  document_id: string
  decision_kind: DecisionKind
  document_number?: string
  document_symbol?: string
  issued_date?: string
  issuing_agency?: string
  excerpt?: string
  effective_from?: string
  effective_to?: string
  status: DecisionStatus
  notes?: string
}

export interface DecisionListFilters {
  q?: string
  document_id?: string
  decision_kind?: DecisionKind | ''
  document_number?: string
  issuing_agency?: string
  status?: DecisionStatus | ''
  issued_date_from?: string
  issued_date_to?: string
  effective_from?: string
  effective_to?: string
  sort_by?:
    | 'created_at'
    | 'updated_at'
    | 'document_number'
    | 'decision_kind'
    | 'issuing_agency'
    | 'status'
    | 'issued_date'
    | 'effective_from'
    | 'effective_to'
  sort_dir?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface DecisionListResponse {
  items: DecisionItem[]
  total: number
  limit: number
  offset: number
}
